File: Filestore/src/main/com/alex/tools/dump/MapPacker.java
package com.alex.tools.dump;

import com.alex.store.Index;
import com.alex.store.Store;
import com.alex.util.XTEAManager;
import com.alex.utils.CompressionUtils;

import java.io.File;
import java.nio.file.Files;

public final class MapPacker {

    public static void pack(Store cache, String path,
                            int regionId, int mapId, int objId) {

        Index mapIndex = cache.getIndexes()[5];

        File map = new File(path, "map_files");
        File obj = new File(path, "land_files");

        if (!map.exists() || !obj.exists()) {
            System.out.println("Missing data.");
            return;
        }

        int x = (regionId >> 8) & 0xFF;
        int y = regionId & 0xFF;

        try {
            File mapData = getFile(map, mapId);
            if (mapData != null) {
                byte[] data = Files.readAllBytes(mapData.toPath());
                if (mapData.getName().endsWith(".gz"))
                    data = CompressionUtils.gunzip(data);
                mapIndex.putFile(mapId, 0, data);
            } else {
                System.out.println("Missing map file: " + mapId);
            }
            File objData = getFile(obj, objId);
            if (objData != null) {
                byte[] data = Files.readAllBytes(objData.toPath());
                if (objData.getName().endsWith(".gz"))
                    data = CompressionUtils.gunzip(data);
                int[] keys = XTEAManager.lookup(regionId);
                mapIndex.putFile(objId, 0, data, keys);
            } else {
                System.out.println("Missing obj file: " + objId);
            }
        } catch (Exception e) {
            System.out.println("Failed region " + regionId + ": " + e.getMessage());
        }
    }

    private static File getFile(File file, int id) {
        File gz = new File(file, id + ".gz");
        if (gz.exists()) return gz;
        File dat = new File(file, id + ".dat");
        if (dat.exists()) return dat;
        return null;
    }
}

File: Filestore/src/main/com/alex/tools/MapPacker.kt
package com.alex.tools

import com.alex.Cache
import com.alex.loaders.MapIndex
import com.alex.utils.Utils
import java.io.File
import java.nio.file.Files

object MapPacker {

    @JvmStatic
    fun pack() {
        val store = Cache.getStore()
        val mapIndex = store.indexes[5]

        val mapDir = File("Maps/map_files/")
        val landDir = File("Maps/land_files/")

        if (!mapDir.exists() || !landDir.exists()) {
            println("Folders not found!")
            return
        }

        val mapFiles = mapDir.listFiles { f -> f.extension == "dat" } ?: return

        var nextArchiveId = mapIndex.lastArchiveId + 1

        for (mapFile in mapFiles) {

            val regionId = mapFile.nameWithoutExtension.toIntOrNull()
            if (regionId == null) {
                println("Invalid map file name: ${mapFile.name}")
                continue
            }

            val landFile = File(landDir, "$regionId.dat")
            if (!landFile.exists()) {
                println("Missing land file for region $regionId")
                continue
            }

            val x = regionId shr 8
            val y = regionId and 0xFF

            val mapName = "m${x}_$y"
            val landName = "l${x}_$y"

            val mapArchiveId = mapIndex.getArchiveId(mapName).takeIf { it != -1 }
                ?: nextArchiveId++

            val landArchiveId = mapIndex.getArchiveId(landName).takeIf { it != -1 }
                ?: nextArchiveId++

            val mapData = Files.readAllBytes(mapFile.toPath())
            val landData = Files.readAllBytes(landFile.toPath())

            if (mapData.isEmpty() || landData.isEmpty()) {
                println("Empty data for region $regionId")
                continue
            }

            mapIndex.putFile(
                mapArchiveId,
                0,
                0,
                mapData,
                null,
                false,
                false,
                Utils.getNameHash(mapName),
                0
            )

            mapIndex.putFile(
                landArchiveId,
                0,
                0,
                landData,
                null,
                false,
                false,
                Utils.getNameHash(landName),
                0
            )

            MapIndex.setRegion(regionId, landArchiveId, mapArchiveId)

            println("Packed region $regionId [$mapName, $landName]")
        }

        mapIndex.rewriteTable()
        mapIndex.resetCachedFiles()

        println("Done packing maps.")
    }
}